File: perf-rust-vs-js/java/Bench.java
// Java side of the cross-language comparison of pure-rand's generators.
//
// Usage: java Bench <generator> [verify | <seeds> <perSeed> <reps>]
// with <generator> one of: xorshift128plus, xoroshiro128plus, congruential32, mersenne.
//
// Same workload as ../rust/src/main.rs and ../js/bench.cjs: for each seed,
// build a generator and produce `perSeed` numbers, folding every value into a
// uint32 checksum all sides must agree on. The 128-bit generators use one
// `long` per 64-bit state word (the JS versions split them into int32 halves);
// mersenne allocates its 624-word state per seed, like the JS version.

import jdk.incubator.vector.IntVector;
import jdk.incubator.vector.VectorOperators;
import jdk.incubator.vector.VectorSpecies;

public final class Bench {
  interface Generator {
    int next();
  }

  // Port of src/generator/xorshift128plus.ts (XorShift128+ with a=23, b=18, c=5).
  static final class XorShift128Plus implements Generator {
    private long s0;
    private long s1;

    XorShift128Plus(int seed) {
      // JS: new XorShift128Plus(-1, ~seed, seed | 0, 0)
      this.s0 = 0xffffffff00000000L | (~seed & 0xffffffffL);
      this.s1 = (seed & 0xffffffffL) << 32;
    }

    public int next() {
      final long a = s0 ^ (s0 << 23);
      final long s1 = this.s1;
      final int out = (int) (s0 + s1); // low 32 bits of s0+s1, like pure-rand
      this.s0 = s1;
      this.s1 = a ^ s1 ^ (a >>> 18) ^ (s1 >>> 5);
      return out;
    }
  }

  // Port of src/generator/xoroshiro128plus.ts (XoroShiro128+ with a=24, b=16, c=37).
  static final class XoroShiro128Plus implements Generator {
    private long s0;
    private long s1;

    XoroShiro128Plus(int seed) {
      // JS: new XoroShiro128Plus(-1, ~seed, seed | 0, 0)
      this.s0 = 0xffffffff00000000L | (~seed & 0xffffffffL);
      this.s1 = (seed & 0xffffffffL) << 32;
    }

    public int next() {
      final int out = (int) (s0 + s1);
      final long a = s0 ^ s1;
      this.s0 = Long.rotateLeft(s0, 24) ^ a ^ (a << 16);
      this.s1 = Long.rotateLeft(a, 37);
      return out;
    }
  }

  // Port of src/generator/congruential32.ts: three LCG steps per next,
  // 15 usable bits taken from each, recombined into one 32-bit output.
  static final class LinearCongruential32 implements Generator {
    private static final int MULTIPLIER = 0x000343fd;
    private static final int INCREMENT = 0x00269ec3;
    // JS MASK_2 = -2147483649 goes through ToInt32 and becomes 0x7fffffff.
    private static final int MASK_2 = 0x7fffffff;
    private static final int MULTIPLIER_2 = 0xa9fc6809; // = a^2 mod 2^32
    private static final int INCREMENT_2 = 0x1e278e7a; // = c*(1 + a) mod 2^32
    private static final int MULTIPLIER_3 = 0x45c82be5; // = a^3 mod 2^32
    private static final int INCREMENT_3 = 0xd2f65b55; // = c*(1 + a + a^2) mod 2^32

    private int seed;

    LinearCongruential32(int seed) {
      this.seed = seed;
    }

    public int next() {
      final int s0 = this.seed;
      final int s1 = s0 * MULTIPLIER + INCREMENT;
      final int s2 = s0 * MULTIPLIER_2 + INCREMENT_2;
      final int s3 = s0 * MULTIPLIER_3 + INCREMENT_3;
      this.seed = s3;
      // (s & 0x7fffffff) >> 16 == (s << 1) >>> 17: drop the sign bit, keep
      // bits 16..30 — shorter encodings than 4-byte mask immediates.
      final int v1 = (s1 << 1) >>> 17;
      final int v2 = (s2 << 1) >>> 17;
      final int v3 = (s3 << 1) >>> 17;
      return v3 | (v2 << 15) | (v1 << 30);
    }
  }

  // Port of src/generator/mersenne.ts (MT19937). Emits the exact same
  // sequence, but refills lazily — when the read index wraps, it twists the
  // whole 624-word block in three tight loops and tempers every word into an
  // output buffer (both vectorizable) — instead of one twist per next() like
  // the JS version. next() is then a plain buffer read; the tempered outputs
  // are identical either way.
  static final class MersenneTwister implements Generator {
    private static final int N = 624;
    private static final int M = 397;
    private static final int A = 0x9908b0df;
    private static final int F = 1812433253;
    private static final int B = 0x9d2c5680;
    private static final int C = 0xefc60000;
    private static final int MASK_LOWER = 0x7fffffff;
    private static final int MASK_UPPER = 0x80000000;

    private final int[] states;
    private final int[] tempered;
    private int index;

    MersenneTwister(int seed) {
      final int[] states = new int[N];
      int prev = seed;
      states[0] = prev;
      for (int idx = 1; idx < N; ++idx) {
        final int xored = prev ^ (prev >>> 30);
        prev = F * xored + idx;
        states[idx] = prev;
      }
      twistBlock(states);
      this.states = states;
      this.tempered = new int[N];
      temperBlock(states, this.tempered);
      this.index = 0;
    }

    private static final VectorSpecies<Integer> SP = IntVector.SPECIES_PREFERRED;

    private static int twistWord(int current, int next, int far) {
      final int y = (current & MASK_UPPER) | (next & MASK_LOWER);
      return far ^ (y >>> 1) ^ (-(y & 1) & A);
    }

    // Vectorized twist of mt[from..to) with the "far" word at idx+farOff.
    // Safe to vectorize: within a vector step the loads (idx+1, idx+farOff)
    // happen before the store to idx, and across steps the store range never
    // overtakes a load range (farOff is +M ahead, or M-N = -227 behind, both
    // beyond the vector length).
    private static void twistRange(int[] mt, int from, int to, int farOff) {
      int idx = from;
      for (int upper = from + SP.loopBound(to - from); idx < upper; idx += SP.length()) {
        final IntVector cur = IntVector.fromArray(SP, mt, idx);
        final IntVector nxt = IntVector.fromArray(SP, mt, idx + 1);
        final IntVector far = IntVector.fromArray(SP, mt, idx + farOff);
        final IntVector y = cur.and(MASK_UPPER).or(nxt.and(MASK_LOWER));
        far.lanewise(VectorOperators.XOR, y.lanewise(VectorOperators.LSHR, 1))
            .lanewise(VectorOperators.XOR, y.and(1).lanewise(VectorOperators.NEG).and(A))
            .intoArray(mt, idx);
      }
      for (; idx < to; ++idx) {
        mt[idx] = twistWord(mt[idx], mt[idx + 1], mt[idx + farOff]);
      }
    }

    private static void twistBlock(int[] mt) {
      twistRange(mt, 0, N - M, M);
      twistRange(mt, N - M, N - 1, M - N);
      mt[N - 1] = twistWord(mt[N - 1], mt[0], mt[M - 1]);
    }

    private static void temperBlock(int[] mt, int[] out) {
      int idx = 0;
      for (int upper = SP.loopBound(N); idx < upper; idx += SP.length()) {
        IntVector y = IntVector.fromArray(SP, mt, idx);
        y = y.lanewise(VectorOperators.XOR, y.lanewise(VectorOperators.LSHR, 11));
        y = y.lanewise(VectorOperators.XOR, y.lanewise(VectorOperators.LSHL, 7).and(B));
        y = y.lanewise(VectorOperators.XOR, y.lanewise(VectorOperators.LSHL, 15).and(C));
        y = y.lanewise(VectorOperators.XOR, y.lanewise(VectorOperators.LSHR, 18));
        y.intoArray(out, idx);
      }
      for (; idx < N; ++idx) {
        int y = mt[idx];
        y ^= y >>> 11;
        y ^= (y << 7) & B;
        y ^= (y << 15) & C;
        y ^= y >>> 18;
        out[idx] = y;
      }
    }

    public int next() {
      if (index == N) {
        twistBlock(states);
        temperBlock(states, tempered);
        index = 0;
      }
      return tempered[index++];
    }
  }

  interface Factory {
    Generator create(int seed);
  }

  // Same seed derivation as the other benches: Knuth multiplicative hash of the index.
  private static int seedAt(int i) {
    return i * (int) 2654435761L;
  }

  private static int run(Factory factory, int numSeeds, int perSeed) {
    int checksum = 0;
    for (int i = 0; i < numSeeds; ++i) {
      final Generator rng = factory.create(seedAt(i));
      for (int j = 0; j < perSeed; ++j) {
        checksum += rng.next();
      }
    }
    return checksum;
  }

  public static void main(String[] args) {
    final String name = args.length > 0 ? args[0] : "xorshift128plus";
    final Factory factory =
        switch (name) {
          case "xorshift128plus" -> XorShift128Plus::new;
          case "xoroshiro128plus" -> XoroShiro128Plus::new;
          case "congruential32" -> LinearCongruential32::new;
          case "mersenne" -> MersenneTwister::new;
          default -> {
            System.err.println("Unknown generator '" + name + "'");
            System.exit(1);
            yield null;
          }
        };

    if (args.length > 1 && args[1].equals("verify")) {
      for (final int seed : new int[] {0, 42, -1, 123456789, -987654321}) {
        final Generator rng = factory.create(seed);
        final StringBuilder values = new StringBuilder();
        for (int i = 0; i != 10; ++i) {
          if (i != 0) values.append(',');
          values.append(rng.next());
        }
        System.out.println("seed=" + seed + " -> " + values);
      }
      return;
    }

    final int numSeeds = args.length > 1 ? Integer.parseInt(args[1]) : 100000;
    final int perSeed = args.length > 2 ? Integer.parseInt(args[2]) : 1000;
    final int reps = args.length > 3 ? Integer.parseInt(args[3]) : 7;

    // Warmup so the JIT reaches steady state before we measure.
    int sink = 0;
    for (int i = 0; i != 3; ++i) sink ^= run(factory, numSeeds, perSeed);
    final int checksum = run(factory, numSeeds, perSeed);
    System.out.println(
        "java "
            + name
            + " | seeds="
            + numSeeds
            + " per_seed="
            + perSeed
            + " checksum="
            + Integer.toUnsignedString(checksum));

    final double[] timingsMs = new double[reps];
    for (int r = 0; r != reps; ++r) {
      final long start = System.nanoTime();
      sink ^= run(factory, numSeeds, perSeed);
      timingsMs[r] = (System.nanoTime() - start) / 1e6;
    }
    if (sink == 123456) System.out.println(); // keep `sink` alive
    java.util.Arrays.sort(timingsMs);
    final double best = timingsMs[0];
    final double median = timingsMs[reps / 2];
    final double total = (double) numSeeds * perSeed;
    System.out.printf(
        "best: %.2f ms | median: %.2f ms | %.3f ns/number | %.0f numbers/sec | %.3f µs per %d-number sequence%n",
        best, median, best * 1e6 / total, total / (best / 1e3), best * 1e3 / numSeeds, perSeed);
  }
}

File: perf-rust-vs-js/python/bench.py
# Pure-Python side of the cross-language comparison of pure-rand's generators.
# No third-party libraries: plain interpreter arithmetic with explicit 32/64-bit
# masking (Python ints are arbitrary-precision, so every step masks back down).
#
# Usage: python3 bench.py <generator> [verify | <seeds> <per_seed> <reps>]
# with <generator> one of: xorshift128plus, xoroshiro128plus, congruential32, mersenne.
#
# Same workload as ../rust/src/main.rs and ../js/bench.cjs: for each seed,
# build a generator and produce `per_seed` numbers, folding every value into a
# uint32 checksum all sides must agree on. State is kept as unsigned ints;
# outputs are converted to signed int32 for `verify` (checksums are unaffected:
# signed and unsigned representations are congruent mod 2^32).

import sys
import time

MASK32 = 0xFFFFFFFF
MASK64 = 0xFFFFFFFFFFFFFFFF


def to_signed32(value):
    return value - 0x100000000 if value >= 0x80000000 else value


# Port of src/generator/xorshift128plus.ts (XorShift128+ with a=23, b=18, c=5).
class XorShift128Plus:
    __slots__ = ("s0", "s1")

    def __init__(self, seed):
        # JS: new XorShift128Plus(-1, ~seed, seed | 0, 0)
        seed &= MASK32
        self.s0 = 0xFFFFFFFF00000000 | (seed ^ MASK32)
        self.s1 = seed << 32

    def next(self):
        s0 = self.s0
        s1 = self.s1
        a = (s0 ^ (s0 << 23)) & MASK64
        out = (s0 + s1) & MASK32  # low 32 bits of s0+s1, like pure-rand
        self.s0 = s1
        self.s1 = a ^ s1 ^ (a >> 18) ^ (s1 >> 5)
        return out


# Port of src/generator/xoroshiro128plus.ts (XoroShiro128+ with a=24, b=16, c=37).
class XoroShiro128Plus:
    __slots__ = ("s0", "s1")

    def __init__(self, seed):
        # JS: new XoroShiro128Plus(-1, ~seed, seed | 0, 0)
        seed &= MASK32
        self.s0 = 0xFFFFFFFF00000000 | (seed ^ MASK32)
        self.s1 = seed << 32

    def next(self):
        s0 = self.s0
        s1 = self.s1
        out = (s0 + s1) & MASK32
        a = s0 ^ s1
        self.s0 = (((s0 << 24) | (s0 >> 40)) ^ a ^ (a << 16)) & MASK64  # rotl(s0, 24)
        self.s1 = ((a << 37) | (a >> 27)) & MASK64  # rotl(a, 37)
        return out


# Port of src/generator/congruential32.ts: three LCG steps per next,
# 15 usable bits taken from each, recombined into one 32-bit output.
class LinearCongruential32:
    __slots__ = ("seed",)

    MULTIPLIER = 0x000343FD
    INCREMENT = 0x00269EC3
    # JS MASK_2 = -2147483649 goes through ToInt32 and becomes 0x7fffffff.
    MASK_2 = 0x7FFFFFFF
    MULTIPLIER_2 = 0xA9FC6809  # = a^2 mod 2^32
    INCREMENT_2 = 0x1E278E7A  # = c*(1 + a) mod 2^32
    MULTIPLIER_3 = 0x45C82BE5  # = a^3 mod 2^32
    INCREMENT_3 = 0xD2F65B55  # = c*(1 + a + a^2) mod 2^32

    def __init__(self, seed):
        self.seed = seed & MASK32

    def next(self):
        s0 = self.seed
        s1 = (s0 * 0x000343FD + 0x00269EC3) & MASK32
        s2 = (s0 * 0xA9FC6809 + 0x1E278E7A) & MASK32
        s3 = (s0 * 0x45C82BE5 + 0xD2F65B55) & MASK32
        self.seed = s3
        v1 = (s1 & 0x7FFFFFFF) >> 16
        v2 = (s2 & 0x7FFFFFFF) >> 16
        v3 = (s3 & 0x7FFFFFFF) >> 16
        return (v3 | (v2 << 15) | (v1 << 30)) & MASK32


# Port of src/generator/mersenne.ts (MT19937 with pure-rand's incremental
# twist: each next tempers the current word then twists one word forward).
class MersenneTwister:
    __slots__ = ("states", "index")

    N = 624
    M = 397
    A = 0x9908B0DF
    F = 1812433253

    def __init__(self, seed):
        n = self.N
        states = [0] * n
        states[0] = seed & MASK32
        for idx in range(1, n):
            xored = states[idx - 1] ^ (states[idx - 1] >> 30)
            states[idx] = (1812433253 * xored + idx) & MASK32
        self.states = states
        self.index = 0
        for idx in range(n):
            self._twisted_next(idx)

    def _twisted_next(self, idx):
        mt = self.states
        next_idx = 0 if idx == 623 else idx + 1
        y = (mt[idx] & 0x80000000) | (mt[next_idx] & 0x7FFFFFFF)
        twisted_idx = idx + 397 if idx < 227 else idx - 227
        mt[idx] = mt[twisted_idx] ^ (y >> 1) ^ (0x9908B0DF if y & 1 else 0)
        return next_idx

    def next(self):
        idx = self.index
        y = self.states[idx]
        y ^= y >> 11
        y ^= (y << 7) & 0x9D2C5680
        y ^= (y << 15) & 0xEFC60000
        y ^= y >> 18
        self.index = self._twisted_next(idx)
        return y & MASK32


GENERATORS = {
    "xorshift128plus": XorShift128Plus,
    "xoroshiro128plus": XoroShiro128Plus,
    "congruential32": LinearCongruential32,
    "mersenne": MersenneTwister,
}


def seed_at(i):
    # Same seed derivation as the other benches: Knuth multiplicative hash of the index.
    return (i * 2654435761) & MASK32


def run(make_generator, num_seeds, per_seed):
    checksum = 0
    for i in range(num_seeds):
        rng = make_generator(seed_at(i))
        rng_next = rng.next
        for _ in range(per_seed):
            checksum += rng_next()
    return checksum & MASK32


def main():
    name = sys.argv[1] if len(sys.argv) > 1 else "xorshift128plus"
    make_generator = GENERATORS.get(name)
    if make_generator is None:
        print(f"Unknown generator '{name}': expected {', '.join(GENERATORS)}", file=sys.stderr)
        sys.exit(1)

    if len(sys.argv) > 2 and sys.argv[2] == "verify":
        for seed in (0, 42, -1, 123456789, -987654321):
            rng = make_generator(seed)
            values = ",".join(str(to_signed32(rng.next())) for _ in range(10))
            print(f"seed={seed} -> {values}")
        return

    num_seeds = int(sys.argv[2]) if len(sys.argv) > 2 else 1000
    per_seed = int(sys.argv[3]) if len(sys.argv) > 3 else 1000
    reps = int(sys.argv[4]) if len(sys.argv) > 4 else 7

    checksum = run(make_generator, num_seeds, per_seed)  # warmup + correctness output
    print(f"python {name} | seeds={num_seeds} per_seed={per_seed} checksum={checksum}")

    timings_ms = []
    for _ in range(reps):
        start = time.perf_counter_ns()
        run(make_generator, num_seeds, per_seed)
        timings_ms.append((time.perf_counter_ns() - start) / 1e6)
    timings_ms.sort()
    best = timings_ms[0]
    median = timings_ms[len(timings_ms) // 2]
    total = num_seeds * per_seed
    print(
        f"best: {best:.2f} ms | median: {median:.2f} ms | "
        f"{best * 1e6 / total:.3f} ns/number | {total / (best / 1e3):.0f} numbers/sec | "
        f"{best * 1e3 / num_seeds:.3f} µs per {per_seed}-number sequence"
    )


if __name__ == "__main__":
    main()
